Sweep barrier test can deadlock: unbounded calls inside a forced stall window, release not cleanup-safe

## Summary

`TestSweepBlocksOnTheLockAndThenFindsTheReference` (`internal/dataplane/store/postgres/sweep_integration_test.go`) deliberately stalls a writer inside an open transaction, holding a digest lock, and then performs **unbounded external calls before releasing that stall**. If any of those calls hangs, the writer stays stalled, the sweep stays blocked behind its lock, and nothing breaks the cycle until the package-wide `go test` timeout fires — 20 minutes, failing the whole gate.

Contention exposed this; the defect is in the harness. Filing separately from #306 and #307 rather than folding it in, because those are runner-level (caching, suite concurrency) and this is a test-structure fault that would survive fixing both.

## The window

The test's design is sound and worth preserving — the collision is forced rather than raced for. The problem is what happens *inside* the forced window:

```go
// writer goroutine is now stalled post-promote, holding the digest lock
select {
case <-stall.reached:
case err := <-written: ...
case <-time.After(30 * time.Second): ...   // bounds ONLY this first wait
}

if f.storedVersions(t, digest) == 0 { ... }  // unbounded
waitForLockWait(t, f)                        // per-call unbounded
close(stall.release)                         // the only release, happy path only
```

Two unbounded calls sit between the stall and its release:

1. **`storedVersions`** (`sweep_integration_test.go:81`) calls `f.blob.ListVersions(context.Background(), key)`. No deadline of any kind.
2. **`waitForLockWait`** (`truncation_seam_integration_test.go:241`) has a 30-second wall-clock loop, but each iteration runs `f.pool.QueryRow(context.Background(), ...)`. The deadline governs *iterations*; a single query that never returns never reaches the next deadline check. A nominal 30-second bound that cannot actually fire.

The `time.After(30 * time.Second)` above bounds only the first barrier, which is the one least likely to hang.

## Observed, 2026-08-17

Cold full-suite pre-push run on #286's branch:

```
panic: test timed out after 20m0s
FAIL orchestrator/internal/dataplane/store/postgres 1203.811s

goroutine 17187 [sync.WaitGroup.Wait, 15 minutes]:
  ...TestSweepBlocksOnTheLockAndThenFindsTheReference(...)
      sweep_integration_test.go:316
goroutine 17177 [chan receive, 15 minutes]:
  ...(*Blob).Get -> (*Store).verifyStored -> putNewObject   // the writer, stalled by design
goroutine 17210 [chan receive, 15 minutes]:
  ...(*stallAfterPromote).RoundTrip                          // the stall, awaiting release
```

The main goroutine is parked in `storedVersions` → minio client → `WaitGroup.Wait` for 15 minutes. The writer is correctly stalled. The release never comes because the goroutine that would send it is the one stuck. Deadlock by construction, triggered by a slow object store.

The same test passes standalone in **1.47s**.

## Second fault: release is not cleanup-safe

`close(stall.release)` runs only on the success path. Any `t.Fatal` in either barrier ends the test with the writer goroutine still blocked, still holding a Postgres transaction and the digest lock. That leaked goroutine can then wedge *other* tests in the same package, turning one assertion failure into a cascade whose root cause is invisible.

## Suggested fix

- Give every call inside the forced window an explicit bounded context, rather than `context.Background()`.
- In `waitForLockWait`, derive a per-query context from the loop's deadline so the bound governs the *call*, not just the iteration.
- Release the stall from `t.Cleanup` via `sync.Once`, so the writer is freed on failure paths as well as success. The test then fails with its real assertion message instead of hanging.
- Audit sibling tests that force a window the same way; this pattern is unlikely to be unique to this one.

## Scope note

Discovered from #286's branch. That branch does not modify `internal/dataplane/store/postgres` — but it does add a test to `internal/dataplane/benchmarkimport`, which runs concurrently under `go test ./...`, so it contributed to the load. "Untouched package" is not by itself a claim of independence, and is not being made as one here.

Related: #306 (gate passes on cached results — the retry that pushed #286 had 71 cached packages and so did not re-exercise this) and #307 (local suite parallelism).

--- a/Makefile
+++ b/Makefile
@@ -76,16 +76,29 @@ test: benchmark-test
 #     internal/dataplane/benchmarkimport  86.9s   (1.5s standalone: ~58x)
 #
 # 40m leaves the long pole better than 3x headroom, and matches what the
-# ubuntu-latest recovery job in ci.yml already uses. Headroom here absorbs
-# CONTENTION rather than slow tests: `./...` runs packages concurrently against
-# one shared Postgres and one MinIO, and the inflation factors above are what
-# that costs.
+# ubuntu-latest recovery job in ci.yml already uses.
 #
-# Linux is not slower. The recovery subset of the stack suite takes 261-289s on
-# ubuntu-latest in CI against 233s on this machine, so the macOS figures are a
-# reasonable basis for both. The full suite has not been timed on Linux, because
-# it does not run there yet -- CI's integration job is still the commented-out
-# template at the foot of ci.yml.
+# The inflation column has TWO causes, not one, which is why its range is so
+# wide. Packages built on `planetest` take a database and a bucket per test on
+# the ALREADY-RUNNING dev stack, so under `./...` they compete for one Postgres
+# and one MinIO -- that is benchmarkimport's ~58x. The stack package instead
+# brings up its own throwaway Compose planes per test, so it is not queueing
+# behind anyone for a shared server; its 1.04x is near-zero for that reason and
+# because, as the long pole, little else is still running near its end. Reading
+# 1.04x as "contention is mild" would be the wrong lesson.
+#
+# Linux is somewhat SLOWER, not faster. On the only subset timed on both, the
+# stack recovery tests take 261-289s on ubuntu-latest in CI against 238.3s here:
+# roughly 1.1-1.2x. Applied to the long pole that is ~900s, still under half the
+# timeout, which is what makes deferring the full Linux measurement safe rather
+# than merely convenient.
+#
+# The full suite has NOT been timed on Linux. It does not run there -- CI's
+# integration job is still the commented-out template at the foot of ci.yml --
+# and standing that up means a paid-API surface and a dispatch trigger that
+# cannot be exercised until it exists on the default branch. #306 was amended to
+# defer it until that job legitimately enters CI, rather than leaving an
+# acceptance criterion nobody could satisfy.
 test-integration:
 	@echo "🧪 Running integration tests..."
 	go test -tags=integration -cover -count=1 -timeout=40m ./...

--- a/Makefile
+++ b/Makefile
@@ -57,9 +57,38 @@ test: benchmark-test
 	go test -cover ./...
 
 # Run integration tests only (requires API keys and external services)
+#
+# -count=1 is REQUIRED, not tidiness. These tests depend on mutable external
+# state -- a Postgres cluster, MinIO buckets, containers -- and Go's test cache
+# keys on inputs it can see, which excludes all of it. A cached PASS is a claim
+# about a plane that may since have been reset, migrated or emptied. Measured on
+# #286's branch: a pre-push run that failed was followed by one that passed with
+# 71 packages `(cached)`, including the two heaviest, so the accepting run had
+# executed strictly less than the run it was taken to overrule (#306).
+#
+# -timeout applies PER TEST BINARY, so it is sized by the slowest single
+# package, not by the suite. Measured on macOS (M3) 2026-08-17, from a fully
+# uncached run of this target -- 747s wall clock, zero packages cached:
+#
+#     internal/dataplane/stack           745.1s   <- the long pole
+#     internal/dataplane/store/postgres  292.8s   (187.2s standalone: 1.6x)
+#     tests/integration                  215.8s
+#     internal/dataplane/benchmarkimport  86.9s   (1.5s standalone: ~58x)
+#
+# 40m leaves the long pole better than 3x headroom, and matches what the
+# ubuntu-latest recovery job in ci.yml already uses. Headroom here absorbs
+# CONTENTION rather than slow tests: `./...` runs packages concurrently against
+# one shared Postgres and one MinIO, and the inflation factors above are what
+# that costs.
+#
+# Linux is not slower. The recovery subset of the stack suite takes 261-289s on
+# ubuntu-latest in CI against 233s on this machine, so the macOS figures are a
+# reasonable basis for both. The full suite has not been timed on Linux, because
+# it does not run there yet -- CI's integration job is still the commented-out
+# template at the foot of ci.yml.
 test-integration:
 	@echo "🧪 Running integration tests..."
-	go test -tags=integration -cover -timeout=20m ./...
+	go test -tags=integration -cover -count=1 -timeout=40m ./...
 
 # Run the GCS adapter tests against a REAL Google Cloud Storage bucket.
 #

--- a/internal/dataplane/store/postgres/sweep_integration_test.go
+++ b/internal/dataplane/store/postgres/sweep_integration_test.go
@@ -392,12 +392,18 @@ type stallAfterPromote struct {
 
 // newStallAfterPromote builds a stall and registers its release as CLEANUP.
 //
-// The release must happen on every exit path, not only the successful one. A
-// stalled writer holds an open transaction and a digest lock; if the test fails
-// an assertion between the stall and its release, `t.Fatal` ends the test
-// goroutine while that writer stays blocked forever — leaking a goroutine that
-// holds locks other tests in this package then wait on. One failed assertion
-// becomes a cascade whose root cause is invisible.
+// The release must happen on every exit path, not only the successful one, and
+// the reason is more specific than a leaked goroutine. The stalled writer holds
+// an ACQUIRED POOLED CONNECTION with an open transaction, and `pgxpool.Close()`
+// waits for acquired connections to be released. So a `t.Fatal` between the
+// stall and its release does not merely strand a goroutine that might collide
+// with a later test — it blocks this fixture's own teardown, forever, every
+// time. That is why the failure presents as a package-wide timeout rather than
+// as one test failing.
+//
+// MEASURED: removing this registration and forcing a failure inside the window
+// reproduces `panic: test timed out after 1m30s`; with it, the same assertion
+// fails in 0.63s.
 //
 // Registering the release here rather than remembering it at each call site is
 // the point: the paths that need it most are the ones nobody writes out.

--- a/internal/dataplane/store/postgres/sweep_integration_test.go
+++ b/internal/dataplane/store/postgres/sweep_integration_test.go
@@ -75,13 +75,31 @@ func (f *fixture) unreferencedObject(t *testing.T, body []byte) string {
 	return digest
 }
 
+// barrierTimeout bounds a single external call made while a test is holding a
+// forced window open.
+//
+// Every such call needs a deadline of its own. A test that stalls a writer
+// inside its transaction and then makes an unbounded call has arranged for one
+// slow response to become a deadlock: the writer cannot proceed, whatever is
+// blocked behind its lock cannot proceed, and the goroutine that would release
+// the stall is the one waiting. Nothing breaks that cycle except the package
+// timeout, which fails the whole suite twenty minutes later and names no cause.
+//
+// Generous on purpose. It is not a performance assertion — a loaded object store
+// legitimately takes seconds — it exists so that "slow" cannot become "never".
+const barrierTimeout = 60 * time.Second
+
 // storedVersions counts what the bucket actually holds for a digest,
 // including delete markers. Zero means the storage is genuinely reclaimed
 // rather than merely hidden behind a marker.
+//
+// Bounded, because this is called inside forced windows: see barrierTimeout.
 func (f *fixture) storedVersions(t *testing.T, digest string) int {
 	t.Helper()
 	key := objectKeyFor(f.organizationID, digest)
-	versions, err := f.blob.ListVersions(context.Background(), key)
+	ctx, cancel := context.WithTimeout(context.Background(), barrierTimeout)
+	defer cancel()
+	versions, err := f.blob.ListVersions(ctx, key)
 	if err != nil {
 		t.Fatalf("list versions of %s: %v", key, err)
 	}
@@ -292,11 +310,7 @@ func TestSweepBlocksOnTheLockAndThenFindsTheReference(t *testing.T) {
 	body := []byte("bytes whose attachment row has not committed yet")
 	digest := digestOf(body)
 
-	stall := &stallAfterPromote{
-		key:     objectKeyFor(f.organizationID, digest),
-		reached: make(chan struct{}),
-		release: make(chan struct{}),
-	}
+	stall := newStallAfterPromote(t, objectKeyFor(f.organizationID, digest))
 	writer := f.storeWithTransport(t, stall)
 
 	written := make(chan error, 1)
@@ -306,7 +320,7 @@ func TestSweepBlocksOnTheLockAndThenFindsTheReference(t *testing.T) {
 	case <-stall.reached:
 	case err := <-written:
 		t.Fatalf("the write finished without ever stalling (%v), so no window was held open", err)
-	case <-time.After(30 * time.Second):
+	case <-time.After(barrierTimeout):
 		t.Fatal("the writer never reached its post-promote read-back")
 	}
 
@@ -332,7 +346,7 @@ func TestSweepBlocksOnTheLockAndThenFindsTheReference(t *testing.T) {
 	// instead would make the test pass whether or not the lock was ever
 	// taken.
 	waitForLockWait(t, f)
-	close(stall.release)
+	stall.free()
 
 	if err := <-written; err != nil {
 		t.Fatalf("the stalled write failed: %v", err)
@@ -371,6 +385,38 @@ type stallAfterPromote struct {
 	release  chan struct{}
 	promoted atomic.Bool
 	once     sync.Once
+	// freed guards the release channel so it can be closed from both the test's
+	// success path and its cleanup. See newStallAfterPromote.
+	freed sync.Once
+}
+
+// newStallAfterPromote builds a stall and registers its release as CLEANUP.
+//
+// The release must happen on every exit path, not only the successful one. A
+// stalled writer holds an open transaction and a digest lock; if the test fails
+// an assertion between the stall and its release, `t.Fatal` ends the test
+// goroutine while that writer stays blocked forever — leaking a goroutine that
+// holds locks other tests in this package then wait on. One failed assertion
+// becomes a cascade whose root cause is invisible.
+//
+// Registering the release here rather than remembering it at each call site is
+// the point: the paths that need it most are the ones nobody writes out.
+func newStallAfterPromote(t *testing.T, key string) *stallAfterPromote {
+	t.Helper()
+	stall := &stallAfterPromote{
+		key:     key,
+		reached: make(chan struct{}),
+		release: make(chan struct{}),
+	}
+	t.Cleanup(stall.free)
+	return stall
+}
+
+// free releases the stalled writer, and is safe to call more than once — the
+// test's own success path calls it too, and a cleanup that panicked on a
+// second close would convert a passing test into a failing one.
+func (s *stallAfterPromote) free() {
+	s.freed.Do(func() { close(s.release) })
 }
 
 func (s *stallAfterPromote) RoundTrip(req *http.Request) (*http.Response, error) {

--- a/internal/dataplane/store/postgres/truncation_seam_integration_test.go
+++ b/internal/dataplane/store/postgres/truncation_seam_integration_test.go
@@ -238,14 +238,30 @@ func TestConcurrentTruncationRetriesOnItsOwnSnapshot(t *testing.T) {
 // Polling pg_stat_activity rather than sleeping, because a sleep long
 // enough to be reliable is also long enough to hide the race it was meant
 // to create.
+// The deadline governs the CALL, not merely the iteration.
+//
+// An earlier version wrapped an unbounded `context.Background()` query in a
+// 30-second loop. That bounds how many times the query is retried, which is not
+// the failure mode: one query that never returns never reaches the next deadline
+// check, so the loop's bound cannot fire. This is called while a writer is
+// deliberately stalled holding a lock, so a hang here deadlocks the test until
+// the package timeout.
 func waitForLockWait(t *testing.T, f *fixture) {
 	t.Helper()
-	deadline := time.Now().Add(30 * time.Second)
-	for time.Now().Before(deadline) {
+	ctx, cancel := context.WithTimeout(context.Background(), barrierTimeout)
+	defer cancel()
+
+	for ctx.Err() == nil {
 		var waiting int
-		if err := f.pool.QueryRow(context.Background(), `
+		if err := f.pool.QueryRow(ctx, `
 			SELECT count(*) FROM pg_stat_activity
 			WHERE datname = current_database() AND wait_event_type = 'Lock'`).Scan(&waiting); err != nil {
+			// A deadline reached mid-query is this helper's own timeout, not a
+			// database fault, and reporting it as one would send the next reader
+			// looking at Postgres.
+			if ctx.Err() != nil {
+				break
+			}
 			t.Fatalf("read pg_stat_activity: %v", err)
 		}
 		if waiting > 0 {
@@ -253,5 +269,6 @@ func waitForLockWait(t *testing.T, f *fixture) {
 		}
 		time.Sleep(20 * time.Millisecond)
 	}
-	t.Fatal("no backend ever blocked on a lock, so the collision this test depends on never happened")
+	t.Fatalf("no backend blocked on a lock within %s, so the collision this test depends on never "+
+		"happened", barrierTimeout)
 }

--- a/internal/dataplane/store/postgres/truncation_seam_integration_test.go
+++ b/internal/dataplane/store/postgres/truncation_seam_integration_test.go
@@ -256,10 +256,19 @@ func waitForLockWait(t *testing.T, f *fixture) {
 		if err := f.pool.QueryRow(ctx, `
 			SELECT count(*) FROM pg_stat_activity
 			WHERE datname = current_database() AND wait_event_type = 'Lock'`).Scan(&waiting); err != nil {
-			// A deadline reached mid-query is this helper's own timeout, not a
-			// database fault, and reporting it as one would send the next reader
-			// looking at Postgres.
-			if ctx.Err() != nil {
+			// Attributed by the ERROR, not by ambient deadline state.
+			//
+			// A deadline reached mid-query is this helper's own timeout rather
+			// than a database fault, and reporting it as one would send the next
+			// reader looking at Postgres. But asking `ctx.Err() != nil` here
+			// answers a different question: it is true for ANY failure that
+			// happens to coincide with the deadline expiring, so a genuine pool
+			// or database fault arriving in that same instant would be swallowed
+			// and reported below as "no backend ever blocked on a lock" -- the
+			// one message guaranteed to send someone looking in the wrong place.
+			// The window is narrow; the cost of landing in it is a debugging
+			// session against a healthy database.
+			if errors.Is(err, context.DeadlineExceeded) {
 				break
 			}
 			t.Fatalf("read pg_stat_activity: %v", err)

